src/ast/passes/codegen_llvm.h

```diff
@@ -363,4 +353,67 @@ class CodegenLLVM : public Visitor<CodegenLLVM, ScopedExpr> {
   State state_ = State::INIT;
 };
 
+class CompileContext : public ast::State<"compile-context"> {
+public:
+  CompileContext() : context(std::make_unique<LLVMContext>()) {};
+  std::unique_ptr<LLVMContext> context;
+};
+
+// LLVMInit will create the required LLVM context which can be subsequently
+// shared by other passes. This should always be added, unless an external
+// `LLVMContext` is injected into the pass ahead of time.
+Pass CreateLLVMInitPass();
+
+class CompiledModule : public ast::State<"compiled-module"> {
+public:
+  CompiledModule(std::unique_ptr<llvm::Module> module)
+      : module(std::move(module)) {};
+  std::unique_ptr<llvm::Module> module;
+};
+
+// CompilePass will compile the primary AST, and emit an LLVM module.
+Pass CreateCompilePass(std::optional<std::reference_wrapper<USDTHelper>>
+                           &&usdt_helper = std::nullopt);
+
+// DumpIR may be used to dump the module, requires a compile pass.
+Pass CreateDumpIRPass(std::ostream &out);
+
+// Verification will validate the IR.
+Pass CreateVerifyPass();
+
+// Optimize pass consumes the emitted LLVM module from the compile pass.
+Pass CreateOptimizePass();
+
+class BpfObject : public ast::State<"bpf-object"> {
+public:
+  BpfObject(std::span<char> data) : data(data.begin(), data.end()) {};
+  std::vector<char> data;
+};
+
+// The object pass produces the ELF data for the BPF bytecode. This is required
+// by the Link pass below, and can be dumped.
+Pass CreateObjectPass();
+
+// DumpASM may be used to dump the produced binary.
+Pass CreateDumpASMPass(std::ostream &out);
+
+// Link produces the final output `BpfBytecode` object.
+Pass CreateLinkPass();
+
+// AllCompilePasses returns a `MultiPass` representing all compile passes, in
```

Comment:
nit: there is some minor inconsistency in the comments here as some are prefixed with part of the function name e.g. "DumpASM" and some are not. I tend to think we don't need to repeat the name if it's right above the function.

Reply:
I'm open to whatever, I would just like to be consistent.

I'd like to think that *one day*, we could clean up comments to the extent that documentation gets generated. In that case, it's nice to have full sentences. I typically used the name because docs should be regular, grammatically correct sentences, and they need a subject, e.g. `Returns a multi-pass representing...` is not a sentence :)

Honestly, I think I was doing `Link blah blah` in some weird way to appease your desire to avoid repetition, but still sneak in a full sentence 😆 . I think I just need to make them consistent, e.g. maybe the compile pass becomes "The compile pass ...." ?

Reply:
Personally, I think sentence fragments are fine but if you want full sentences, just include the whole function name for now to be consistent - at least in this file.

Reply:
I stuck with fragments here and made them consistent. I don't want to get hung up on this, but it's nice to have something uniform at some point. (Having automated developer docs in the future would be a good forcing function for that.)